File: sistema.boletas.backend/src/main/java/drej/sistema/boletas/services/BoletaService.java
package drej.sistema.boletas.services;

import drej.sistema.boletas.models.Boleta;
import drej.sistema.boletas.models.Concepto;
import drej.sistema.boletas.models.Persona;
import drej.sistema.boletas.models.record.BoletaDTO;
import drej.sistema.boletas.models.record.ConceptoDTO;
import drej.sistema.boletas.models.record.RegPensionarioDetalleDTO;
import drej.sistema.boletas.repository.BoletaRepository;
import drej.sistema.boletas.repository.PersonaRepository;
import jakarta.transaction.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Optional;

@Service
public class BoletaService {

    private final BoletaRepository boletaRepository;
    private final PersonaRepository personaRepository;

    public BoletaService(BoletaRepository boletaRepository, PersonaRepository personaRepository) {
        this.boletaRepository = boletaRepository;
        this.personaRepository = personaRepository;
    }

    // Subir varias boletas
    @Transactional
    public void guardarBoletas(List<BoletaDTO> lista) {
        for (BoletaDTO dto : lista) {
            guardarBoleta(dto);
        }
    }

    // Guardar 1 boleta
    @Transactional
    public Boleta guardarBoleta(BoletaDTO dto) {
        // Buscar o crear Persona
        Persona persona = personaRepository.findByDocumentoIdentidad(dto.documento_identidad())
                .orElseGet(() -> {
                    Persona p = new Persona();
                    p.setNombres(dto.nombres());
                    p.setApellidos(dto.apellidos());
                    p.setDocumentoIdentidad(dto.documento_identidad());
                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
                    LocalDate fecha = null;
                    String fechaNacimientoStr = dto.fecha_nacimiento();
                    if (fechaNacimientoStr != null && !fechaNacimientoStr.isBlank()) {
                        fecha = LocalDate.parse(fechaNacimientoStr, formatter);
                    }
                    p.setFechaNacimiento(fecha);
                    return personaRepository.save(p);
                });

        // Actualiza campos si hay datos nuevos
        boolean actualizado = false;
        if (dto.nombres() != null && !dto.nombres().isBlank() && !dto.nombres().equals(persona.getNombres())) {
            persona.setNombres(dto.nombres());
            actualizado = true;
        }
        if (dto.apellidos() != null && !dto.apellidos().isBlank() && !dto.apellidos().equals(persona.getApellidos())) {
            persona.setApellidos(dto.apellidos());
            actualizado = true;
        }
        String fechaNacimientoStr = dto.fecha_nacimiento();
        if (fechaNacimientoStr != null && !fechaNacimientoStr.isBlank()) {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
            LocalDate fecha = LocalDate.parse(fechaNacimientoStr, formatter);
            if (persona.getFechaNacimiento() == null || !fecha.equals(persona.getFechaNacimiento())) {
                persona.setFechaNacimiento(fecha);
                actualizado = true;
            }
        }
        if (actualizado) {
            personaRepository.save(persona);
        }

        // Crear boleta y asociar Persona
        Boleta boleta = new Boleta();
        boleta.setArchivoOrigen(dto.archivo_origen());
        boleta.setRawLength(dto.raw_length());
        boleta.setSecuencia(dto.secuencia());
        boleta.setTipoPensionista(dto.tipo_pensionista());
        boleta.setTipoPension(dto.tipo_pension());
        boleta.setLeyendaMensual(dto.leyenda_mensual());
        boleta.setRegimenPensionario(dto.regimen_pensionario());
        boleta.setCodigoEncabezado(dto.codigo_encabezado());
        boleta.setRucBloque(dto.ruc_bloque());
        boleta.setMes(dto.mes());
        boleta.setAnio(dto.anio());
        boleta.setEstado(dto.estado());
        boleta.setEstablecimiento(dto.establecimiento());
        boleta.setCargo(dto.cargo());
        boleta.setTipoServidor(dto.tipo_servidor());
        boleta.setNivelMagHoras(dto.nivel_mag_horas());
        boleta.setTiempoServicio(dto.tiempo_servicio());
        boleta.setLeyendaPermanente(dto.leyenda_permanente());
        boleta.setFechaIngresoRegistro(dto.fecha_ingreso_registro());
        boleta.setFechaTerminoRegistro(dto.fecha_termino_registro());
        boleta.setCuentaPrincipal(dto.cuenta_principal());
        boleta.setCuentasTodas(dto.cuentas_todas());
        boleta.setTotalRemuneraciones(dto.total_remuneraciones());
        boleta.setTotalDescuentos(dto.total_descuentos());
        boleta.setTotalLiquido(dto.total_liquido());
        boleta.setMontoImponible(dto.monto_imponible());
        boleta.setPersona(persona);

        if (dto.conceptos() != null) {
            List<Concepto> conceptos = dto.conceptos().stream()
                    .map(cdto -> {
                        Concepto c = new Concepto();
                        c.setTipo(cdto.tipo());
                        c.setConcepto(cdto.concepto());
                        c.setMonto(cdto.monto());
                        c.setBoleta(boleta);
                        return c;
                    })
                    .toList();
            boleta.setConceptos(conceptos);
        }
        boletaRepository.save(boleta);
        return boleta;
    }

    // Obtener boleta por ID
    public Optional<Boleta> obtenerBoleta(Long id) {
        return boletaRepository.findById(id);
    }

    // Editar boleta
    @Transactional
    public Boleta editarBoleta(Long id, BoletaDTO dto) {
        Boleta boleta = boletaRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Boleta no encontrada"));
        // Actualiza los campos según tus necesidades
        boleta.setEstado(dto.estado());
        boleta.setMes(dto.mes());
        boleta.setAnio(dto.anio());
        // ... otros campos
        // Notas: puedes actualizar más según lo que permita tu lógica
        return boletaRepository.save(boleta);
    }

    // Eliminar boleta
    @Transactional
    public void eliminarBoleta(Long id) {
        boletaRepository.deleteById(id);
    }

    // Exportar boletas (CSV o PDF - implementar según tu caso)
    public byte[] exportarBoletasCSV(List<BoletaDTO> boletas) {
        // Implementar exportación a CSV y retornar byte[]
        return new byte[0];
    }

    // Listar boletas (global)
    public List<BoletaDTO> listarBoletasDTO() {
        return boletaRepository.findAll().stream()
                .map(this::toBoletaDTO)
                .toList();
    }

    // Listar boletas paginado
    public Page<Boleta> listarBoletasPaginado(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return boletaRepository.findAll(pageable);
    }

    // Buscar boletas por filtros (ejemplo por estado, rango fechas, etc.)
    public List<BoletaDTO> buscarBoletasPorEstado(String estado) {
        return boletaRepository.findAll().stream()
                .filter(b -> estado.equals(b.getEstado()))
                .map(this::toBoletaDTO)
                .toList();
    }

    // Obtener boletas por persona ID
    public List<BoletaDTO> obtenerBoletasID(Long personaId) {
        return boletaRepository.findByPersonaId(personaId)
                .stream()
                .map(this::toBoletaDTO)
                .toList();
    }

    // Obtener boletas por persona
    public List<BoletaDTO> obtenerBoletasPersona(Persona persona){
        return boletaRepository.findByPersona(persona)
                .stream()
                .map(this::toBoletaDTO)
                .toList();
    }

    // Convertidor boletas
    public BoletaDTO toBoletaDTO(Boleta boleta) {
        Persona persona = boleta.getPersona();
        return new BoletaDTO(
                boleta.getArchivoOrigen(),
                boleta.getRawLength(),
                boleta.getConceptos().stream()
                        .map(c -> new ConceptoDTO(c.getTipo(), c.getConcepto(), c.getMonto()))
                        .toList(),
                boleta.getSecuencia(),
                boleta.getCodigoEncabezado(),
                boleta.getRucBloque(),
                boleta.getMes(),
                boleta.getAnio(),
                boleta.getEstado(),
                persona.getApellidos(),
                persona.getNombres(),
                persona.getFechaNacimiento() != null ? persona.getFechaNacimiento().toString() : null,
                persona.getDocumentoIdentidad(),
                boleta.getEstablecimiento(),
                boleta.getCargo(),
                boleta.getTipoServidor(),
                boleta.getTipoPensionista(),
                boleta.getTipoPension(),
                boleta.getNivelMagHoras(),
                boleta.getTiempoServicio(),
                boleta.getLeyendaPermanente(),
                boleta.getLeyendaMensual(),
                boleta.getFechaIngresoRegistro(),
                boleta.getFechaTerminoRegistro(),
                boleta.getCuentaPrincipal(),
                boleta.getCuentasTodas(),
                boleta.getRegPensionarioDetalle() == null ? null :
                        new RegPensionarioDetalleDTO(
                                boleta.getRegPensionarioDetalle().getRaw(),
                                boleta.getRegPensionarioDetalle().getAfiliacion()
                        ),
                boleta.getRegimenPensionario(),
                boleta.getTotalRemuneraciones(),
                boleta.getTotalDescuentos(),
                boleta.getTotalLiquido(),
                boleta.getMontoImponible()
        );
    }

    // Listar todas las boletas (entidad)
    public List<Boleta> listarBoletas() {
        return boletaRepository.findAll();
    }
}

File: backend-boletas-laravel/app/Services/BoletaService.php
<?php

namespace App\Services;

use App\Models\Boleta;
use App\Models\Concepto;
use App\Models\Persona;
use Illuminate\Support\Facades\DB;

class BoletaService
{
    // Subir varias boletas (DTOs)
    public function guardarBoletas(array $lista)
    {
        DB::transaction(function() use ($lista) {
            foreach ($lista as $dto) {
                $this->guardarBoleta($dto);
            }
        });
    }

    // Guardar una boleta (DTO in, BoletaDTO out)
    public function guardarBoleta(array $dto)
    {
        // VALIDACIÓN: documento_identidad es obligatorio
        if (empty($dto['documento_identidad'])) {
            throw new \InvalidArgumentException('El campo documento_identidad es obligatorio');
        }

        // Buscar o crear Persona
        $fechaNac = $dto['fecha_nacimiento'] ?? null;

        $persona = Persona::firstOrCreate(
            ['documento_identidad' => $dto['documento_identidad']],
            [
                'nombres' => $dto['nombres'] ?? '',
                'apellidos' => $dto['apellidos'] ?? '',
                'fecha_nacimiento' => $this->parseFecha($fechaNac),
            ]
        );

        $persona->fill([
            'nombres' => $dto['nombres'] ?? $persona->nombres,
            'apellidos' => $dto['apellidos'] ?? $persona->apellidos,
            'fecha_nacimiento' => $this->parseFecha($fechaNac) ?? $persona->fecha_nacimiento,
        ]);
        $persona->save();

        // Crear boleta
        $boletaData = [
            'archivo_origen' => $dto['archivo_origen'] ?? null,
            'raw_length' => $dto['raw_length'] ?? null,
            'secuencia' => $dto['secuencia'] ?? null,
            'codigo_encabezado' => $dto['codigo_encabezado'] ?? null,
            'ruc_bloque' => $dto['ruc_bloque'] ?? null,
            'mes' => $dto['mes'] ?? null,
            'anio' => $dto['anio'] ?? null,
            'estado' => $dto['estado'] ?? null,
            'establecimiento' => $dto['establecimiento'] ?? null,
            'cargo' => $dto['cargo'] ?? null,
            'tipo_servidor' => $dto['tipo_servidor'] ?? null,
            'tipo_pensionista' => $dto['tipo_pensionista'] ?? null,
            'tipo_pension' => $dto['tipo_pension'] ?? null,
            'nivel_mag_horas' => $dto['nivel_mag_horas'] ?? null,
            'tiempo_servicio' => $dto['tiempo_servicio'] ?? null,
            'leyenda_permanente' => $dto['leyenda_permanente'] ?? null,
            'leyenda_mensual' => $dto['leyenda_mensual'] ?? null,
            'fecha_ingreso_registro' => $dto['fecha_ingreso_registro'] ?? null,
            'fecha_termino_registro' => $dto['fecha_termino_registro'] ?? null,
            'cuenta_principal' => $dto['cuenta_principal'] ?? null,
            'cuentas_todas' => $dto['cuentas_todas'] ?? [],
            'reg_pensionario_detalle' => $dto['reg_pensionario_detalle'] ?? null,
            'regimen_pensionario' => $dto['regimen_pensionario'] ?? null,
            'total_remuneraciones' => $dto['total_remuneraciones'] ?? null,
            'total_descuentos' => $dto['total_descuentos'] ?? null,
            'total_liquido' => $dto['total_liquido'] ?? null,
            'monto_imponible' => $dto['monto_imponible'] ?? null,
            'persona_id' => $persona->id,
        ];

        $boleta = new Boleta($boletaData);
        $boleta->save();

        // Conceptos
        if (isset($dto['conceptos']) && is_array($dto['conceptos'])) {
            foreach ($dto['conceptos'] as $c) {
                $concepto = new Concepto([
                    'tipo' => $c['tipo'] ?? '',
                    'concepto' => $c['concepto'] ?? '',
                    'monto' => $c['monto'] ?? 0,
                ]);
                $concepto->boleta()->associate($boleta);
                $concepto->save();
            }
        }

        return $this->toBoletaDTO($boleta->fresh(['persona', 'conceptos']));
    }

    // Obtener boleta por ID (BoletaDTO)
    public function obtenerBoleta($id)
    {
        $boleta = Boleta::with(['persona', 'conceptos'])->find($id);
        return $boleta ? $this->toBoletaDTO($boleta) : null;
    }

    // Editar boleta (BoletaDTO in, BoletaDTO out)
    public function editarBoleta($id, array $dto)
    {
        $boleta = Boleta::findOrFail($id);
        $boleta->fill($dto);
        $boleta->save();
        return $this->toBoletaDTO($boleta->fresh(['persona', 'conceptos']));
    }

    // Eliminar boleta
    public function eliminarBoleta($id)
    {
        Boleta::destroy($id);
    }

    // Listar boletas paginadas (formato Spring, BoletaDTOs)
    public function listarBoletasPaginado($page = 0, $size = 30)
    {
        $paginator = Boleta::with(['persona', 'conceptos'])->paginate($size, ['*'], 'page', $page + 1);
        return [
            'content' => $paginator->getCollection()->map(fn($b) => $this->toBoletaDTO($b))->all(),
            'totalElements' => $paginator->total(),
            'totalPages' => $paginator->lastPage(),
            'size' => $paginator->perPage(),
            'number' => $paginator->currentPage() - 1,
        ];
    }

    // Obtener boletas por persona_id (BoletaDTOs)
    public function obtenerBoletasPorPersona($personaId)
    {
        return Boleta::where('persona_id', $personaId)->with('conceptos')->get()->map(fn($b) => $this->toBoletaDTO($b));
    }

    // Listar todas las boletas (BoletaDTOs)
    public function listarBoletas()
    {
        return Boleta::with(['persona', 'conceptos'])->get()->map(fn($b) => $this->toBoletaDTO($b));
    }

    // Convertir a DTO (idéntico a record BoletaDTO)
    public function toBoletaDTO($boleta)
    {
        $persona = $boleta->persona;
        return [
            'id' => $boleta->id,
            'archivo_origen' => $boleta->archivo_origen,
            'raw_length' => $boleta->raw_length,
            'conceptos' => $boleta->conceptos->map(fn($c) => [
                'tipo' => $c->tipo,
                'concepto' => $c->concepto,
                'monto' => $c->monto,
            ])->all(),
            'secuencia' => $boleta->secuencia,
            'codigo_encabezado' => $boleta->codigo_encabezado,
            'ruc_bloque' => $boleta->ruc_bloque,
            'mes' => $boleta->mes,
            'anio' => $boleta->anio,
            'estado' => $boleta->estado,
            'apellidos' => $persona?->apellidos,
            'nombres' => $persona?->nombres,
            'fecha_nacimiento' => $persona?->fecha_nacimiento ? (string)$persona->fecha_nacimiento : null,
            'documento_identidad' => $persona?->documento_identidad,
            'establecimiento' => $boleta->establecimiento,
            'cargo' => $boleta->cargo,
            'tipo_servidor' => $boleta->tipo_servidor,
            'tipo_pensionista' => $boleta->tipo_pensionista,
            'tipo_pension' => $boleta->tipo_pension,
            'nivel_mag_horas' => $boleta->nivel_mag_horas,
            'tiempo_servicio' => $boleta->tiempo_servicio,
            'leyenda_permanente' => $boleta->leyenda_permanente,
            'leyenda_mensual' => $boleta->leyenda_mensual,
            'fecha_ingreso_registro' => $boleta->fecha_ingreso_registro,
            'fecha_termino_registro' => $boleta->fecha_termino_registro,
            'cuenta_principal' => $boleta->cuenta_principal,
            'cuentas_todas' => $boleta->cuentas_todas,
            'reg_pensionario_detalle' => $boleta->reg_pensionario_detalle,
            'regimen_pensionario' => $boleta->regimen_pensionario,
            'total_remuneraciones' => $boleta->total_remuneraciones,
            'total_descuentos' => $boleta->total_descuentos,
            'total_liquido' => $boleta->total_liquido,
            'monto_imponible' => $boleta->monto_imponible,
        ];
    }

    private function parseFecha($fecha)
    {
        if (empty($fecha)) return null;
        if (strpos($fecha, '/') !== false) {
            // dd/MM/yyyy → yyyy-MM-dd
            $partes = explode('/', $fecha);
            if (count($partes) === 3) {
                return $partes[2] . '-' . $partes[1] . '-' . $partes[0];
            }
        }
        // Si ya viene yyyy-MM-dd, la retorna tal cual
        if (preg_match('/^\d{4}-\d{2}-\d{2}$/', $fecha)) {
            return $fecha;
        }
        return null;
    }
}